### ISSUE TITLE
go/types, types2: nil underlying for selectors on the current type decl

### ISSUE THREAD

**Issue by findleyr** [golang/go#51509]
The following code crashes the type checker with a "nil underlying" panic, because we try to check the underlying when selection fails:

```
type X X.t
```
https://go.dev/play/p/H3Dv94uxQL9?v=gotip

Marking as a release blocker as this is a crash.

CC @griesemer

**Comment by griesemer** [comment]
It's notoriously difficult to prevent crashes in a compiler. We should fix this of course, but I think we'd be ok even if this made it into the release as this is pretty pathological.

**Comment by ianlancetaylor** [comment]
Yes: a compiler crash should be fixed, of course, but in general a compiler crash on invalid code is not a release blocker.  It's just a really really bad error message.

**Comment by findleyr** [comment]
Agree that it may not be a release blocker, though it also crashes gopls in which case it is more than an error message.

One argument for the severity / commonality of this pattern is that it was introduced in https://go.dev/cl/378175, i.e. post beta, and seems to have been encountered pretty quickly inside of Google.
